### ISSUE TITLE
Karma PhantomJS 2 issue with SystemJS

### ISSUE THREAD

**Issue by OClement** [systemjs/systemjs#1081]
I am experiencing an issue I don't quite understand yet when running tests through PhantomJs
For the records, everything works fine when running through Chrome

I think I traced the issue back to SystemJs;

---

I am using Typescript 1.8 beta to generate JS using System Module Bundles for the application code (Angular2), and discrete JS files for Specs.

So following the various examples on how to setup Karma for System / Angular2, I created the following __[shim](http://pastebin.com/1H5EXqKB)__ that is loaded by Karma and is responsible of loading Spec Files through SystemJS

This is the __[stacktrace](http://pastebin.com/UYiY1Z0S)__ I get

Here is the generated JS file it is trying to load:
```
System.register(['angular2/testing'], function(exports_1) {
    "use strict";
    var testing_1;
    function main() {
        testing_1.describe("IssueRepositoryService", function () {
            testing_1.it('should Fail', function () {
                testing_1.expect(true).toBe(false);
            });
        });
    }
    exports_1("main", main);
    return {
        setters:[
            function (testing_1_1) {
                testing_1 = testing_1_1;
            }],
        execute: function() {
        }
    }
});
```

---

At some point, when it loads the module, it calls `doDynamicExecute` and fails at the try/catch statement (`load.execute();` line 715 in system.src.js version 0.19.6)

`load.excute` in turn calls `loader.newModule(entry.declarative ? entry.module.exports : entry.esModule)`

In my case, `entry.declarative == true`, but `entry.module == null`

Not sure what it's all about :/

Thoughts? Does that seems like a legit bug or is that a config error or something else?

**Comment by guybedford** [comment]
It would help to have some more info here. What is the `load.name` of the load that is failing? What does the dependency tree look like? Is this the only System.import statement in the app, or are there others happening?

**Comment by OClement** [comment]
`load.name`: "http://localhost:9876/base/Tests/Modules/TWC.IssueTrak.SPA.Web.Issue.Tests/Tests/webDAL/issue.repository.service.js"
Which is basically the path of the spec file being loaded (file definitely exist at this path and can be loaded by the browser by navigating there)

So far it is the only spec file being loaded from the Shim configuration making calls to System.import

The other files loaded are served through Karma as per usual;
Not sure how I can express how the dependency tree looks like;

Relevant part of the `karma.conf.js` file:

```
files: [
      {
        pattern: 'node_modules/angular2/bundles/angular2-polyfills.js',
        included: true,
        watched: false
      },
      {
        pattern: 'node_modules/systemjs/dist/system.src.js',
        included: true,
        watched: false
      },
      { pattern: 'node_modules/systemjs/dist/*.js', included: false, watched: false },

      {
        pattern: 'node_modules/rxjs/bundles/Rx.js',
        included: true,
        watched: false
      },
      {
        pattern: 'node_modules/angular2/bundles/angular2.dev.js',
        included: true,
        watched: false,
        served: true
      },
      {
        pattern: 'node_modules/angular2/bundles/http.dev.js',
        included: true,
        watched: false,
        served: true
      },
      {
        pattern: 'node_modules/angular2/bundles/router.dev.js',
        included: true,
        watched: false,
        served: true
      },
      {
        pattern: 'node_modules/angular2/bundles/testing.dev.js',
        included: true,
        watched: false,
        served: true
      },
      {
        pattern: 'karma-test-shim.js',
        included: true,
        watched: true
      },

      // ************************** Bringing attention ************//
      // paths loaded via module imports (this is the application's bundles, consisting of multiple System.register each)
      // ******************
      {
        pattern: path.relative(__dirname, conf.bundlesDest.concat('/*.js')),
        included: true,
        watched: true,
        serve: true
      },
      {
        pattern: 'Tests/**/*.js',
        watched: true,
        included: false,
        serve: true
      },
      {
        pattern: 'Tests/**/*.ts', //Serve ts source files for sourcemaps in dev tools
        included: false,
        watched: true
      },
    ],
    exclude: [
      '*.d.ts',
      '**/TestResults/**/*.js'
    ],
```

----

In short, the application consists of multiple `projects`, each consists of multiple TS files that are bundled in one SystemJs Package by Typescript 1.8; It is very similar to what SystemJs-Builder output, with some minor differences.

These bundles are either loaded upfront through <script> tags, or lazy loaded (Async Routes in Ng2) using System.import. You can think of a bundle as being a package with everything a certain Route/View needs.
There can be inter-dependencies (Module 1 of Bundle A import Module 2 of Bundle B for instance)

However, for testing purposes, all bundles are loaded by Karma right away.
Only the spec files are loaded asyncronously, and only when they all are loaded is Karma started with `karma.start()`

Let me know if you require more information

**Comment by guybedford** [comment]
The pastebin stack you posted doesn't actually include the error message itself. Can you clarify the actual error message again?

**Comment by OClement** [comment]
Well that's the problem
I only have the stack, no message whatsoever

The only extra info I can extract is this:

```
PhantomJS 2.1.1 (Windows 8 0.0.0) LOG: 'line', '::', 740

PhantomJS 2.1.1 (Windows 8 0.0.0) LOG: 'sourceURL', '::', 'http://localhost:9876/base/node_modules/systemjs/dist/system.src.js'
```

**Comment by guybedford** [comment]
There's always a message! If you add a catch handler that logs `err.message` for example?

**Comment by OClement** [comment]
Wellll.... no, not in this case :P

I litterally itterate over the error object to make sure I log every single keys/values and all I have is `stack` (pasteBin link in original post), `line` and `sourceURL` both shown in my latest post

I also tried catching the error at both the `System.import` statement and the `Promise.all` one

**Comment by guybedford** [comment]
If you're sure that's the start of the stack then it looks like the error is in http://localhost:9876/base/node_modules/angular2/bundles/angular2.dev.js 786:25 surely?

**Comment by OClement** [comment]
However, looks like I didn't do all my homeworks properly;

I just tried removing all import statements from the spec file, and it works

So it is related to the spec file having import statements; If I try for instance to `import {Component} from 'angular2/core'` then log Component to make sure it is used somehwere in the code, and then it will as per the description above

**Comment by guybedford** [comment]
From the stack you provided, it is an execution on line 786 of angular2.dev.js?

**Comment by OClement** [comment]
Yes it looks like so indeed - so you think it might be an NG2 issue maybe?
